bundles/org.openhab.core.config.discovery.usbserial.ser2net/src/main/java/org/openhab/core/config/discovery/usbserial/ser2net/internal/Ser2NetUsbSerialDiscovery.java

```diff
@@ -0,0 +1,206 @@
+
+    static final String PROPERTY_INTERFACE_NUMBER = "bInterfaceNumber";
+    static final String PROPERTY_INTERFACE = "interface";
+
+    static final String SER2NET = "ser2net";
+    static final String SERIALUSB = "serialusb";
+    static final String TELNET_RFC2217_TCP = "telnet(rfc2217),tcp";
+
+    static final Duration SINGLE_SCAN_DURATION = Duration.ofSeconds(5);
+    static final String SERIAL_PORT_NAME_FORMAT = "rfc2217://%s:%s";
+
+    private final Set<UsbSerialDiscoveryListener> discoveryListeners = new CopyOnWriteArraySet<>();
+    private final MDNSClient mdnsClient;
+
+    private boolean backgroundDiscoveryEnabled = false;
+
+    private Set<UsbSerialDeviceInformation> lastScanResult = new HashSet<>();
+
+    @Activate
+    public Ser2NetUsbSerialDiscovery(final @Reference MDNSClient mdnsClient) {
+        this.mdnsClient = mdnsClient;
+    }
+
+    @Override
+    public void registerDiscoveryListener(UsbSerialDiscoveryListener listener) {
+        discoveryListeners.add(listener);
+    }
+
+    @Override
+    public void unregisterDiscoveryListener(UsbSerialDiscoveryListener listener) {
+        discoveryListeners.remove(listener);
+    }
+
+    @Override
+    public synchronized void startBackgroundScanning() {
+        backgroundDiscoveryEnabled = true;
+        mdnsClient.addServiceListener(SERVICE_TYPE, this);
+        logger.debug("Started ser2net USB-Serial mDNS background discovery");
+    }
+
+    @Override
+    public synchronized void stopBackgroundScanning() {
+        backgroundDiscoveryEnabled = false;
+        mdnsClient.removeServiceListener(SERVICE_TYPE, this);
+        logger.debug("Stopped ser2net USB-Serial mDNS background discovery");
+    }
+
+    @Override
+    public synchronized void doSingleScan() {
+        logger.debug("Starting ser2net USB-Serial mDNS single discovery scan");
+
+        Set<UsbSerialDeviceInformation> scanResult = Stream.of(mdnsClient.list(SERVICE_TYPE, SINGLE_SCAN_DURATION))
+                .map(this::createUsbSerialDeviceInformation) //
+                .filter(Optional::isPresent) //
+                .map(Optional::get) //
+                .collect(Collectors.toSet());
+
+        Set<UsbSerialDeviceInformation> added = setDifference(scanResult, lastScanResult);
+        Set<UsbSerialDeviceInformation> removed = setDifference(lastScanResult, scanResult);
+        Set<UsbSerialDeviceInformation> unchanged = setDifference(scanResult, added);
+
+        lastScanResult = scanResult;
+
+        removed.stream().forEach(this::announceRemovedDevice);
+        added.stream().forEach(this::announceAddedDevice);
+        unchanged.stream().forEach(this::announceAddedDevice);
+
+        logger.debug("Completed ser2net USB-Serial mDNS single discovery scan");
+    }
+
+    private <T> Set<T> setDifference(Set<T> set1, Set<T> set2) {
+        Set<T> result = new HashSet<>(set1);
+        result.removeAll(set2);
+        return Set.copyOf(result);
```

Comment:
Do you really need to return a copy here? I think you already created a copy of the set which will be modified and the set entries are immutable at all.

Reply:
Yes it was a bit overkill indeed. It was a leftover inspired by `DeltaUsbSerialScanner` code.